File: src/main/java/com/alten/springboot/taskmanagerclient/controller/TaskManagerClientController.java
package com.alten.springboot.taskmanagerclient.controller;

import com.alten.springboot.taskmanagerclient.model.EmployeeDto;
import com.alten.springboot.taskmanagerclient.model.RandomPopulationInputDto;
import com.alten.springboot.taskmanagerclient.model.TaskDto;
import com.alten.springboot.taskmanagerclient.model.TeamDto;
import com.alten.springboot.taskmanagerclient.service.EmployeeService;
import com.alten.springboot.taskmanagerclient.service.LoginService;
import com.alten.springboot.taskmanagerclient.service.TaskService;
import com.alten.springboot.taskmanagerclient.service.TeamService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpHeaders;
import org.springframework.stereotype.Controller;
import org.springframework.ui.Model;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@Controller
public class TaskManagerClientController {

    private static CurrentSessionInfo session;
    @Autowired
    private EmployeeService employeeService;
    @Autowired
    private TaskService taskService;
    @Autowired
    private LoginService loginService;
    @Autowired
    private TeamService teamService;

    @GetMapping("/home")
    public String sayHello(Model theModel) {

        if (session == null) {
            return "redirect:/showLoginForm";
        }

        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("admin", session.isAdmin());

        return "Home";
    }

    @GetMapping("/employees")
    public String getEmployees(Model theModel) {

        if (session == null) {
            return "redirect:/showLoginForm";
        }

        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("employees", employeeService.getEmployees());
        theModel.addAttribute("admin", session.isAdmin());
        return "Employees";
    }

    @GetMapping("/tasks")
    public String getTasks(Model theModel) {

        if (session == null) {
            return "redirect:/showLoginForm";
        }
        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("tasks", taskService.getTasks());
        theModel.addAttribute("admin", session.isAdmin());

        return "Tasks";
    }

    @GetMapping("/teams")
    public String getTeams(Model theModel) {

        if (session == null) {
            return "redirect:/showLoginForm";
        }

        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("teams", teamService.getTeams());
        theModel.addAttribute("admin", session.isAdmin());
        return "Teams";
    }

    @GetMapping("/tasks/{employeeId}")
    public String getTasks(@PathVariable("employeeId") int employeeId, Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }

        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("tasksOwner", employeeService.getEmployee(employeeId));
        theModel.addAttribute("tasks", taskService.getTasksByEmployeeId(employeeId));
        theModel.addAttribute("admin", session.isAdmin());
        return "TasksByEmployee";
    }

    @GetMapping("/teams/{teamId}")
    public String getEmployeesByTeam(@PathVariable("teamId") int teamId, Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }

        TeamDto theTeam = teamService.getTeam(teamId);

        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("team", theTeam);
        theModel.addAttribute("employees", theTeam.getEmployees());
        theModel.addAttribute("admin", session.isAdmin());
        return "EmployeesByTeam";
    }

    @GetMapping("/showLoginForm")
    public String showLoginForm(Model theModel) {

        return "Login";
    }

    @PostMapping("/login")
    public String login(@RequestParam("username") String username, @RequestParam("password") String password,
                        Model theModel) {
        HttpHeaders headers = new HttpHeaders();

        EmployeeDto result = loginService.login(username, password, headers);
        if (result != null) {

            session = new CurrentSessionInfo(headers, result);

            return "redirect:/home";
        } else {
            theModel.addAttribute("error", true);
            return showLoginForm(theModel);
        }

    }

    @GetMapping("/logout")
    public String logout(Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }

        loginService.logout();
        session = null;
        theModel.addAttribute("logout", true);
        return showLoginForm(theModel);
    }

    @GetMapping("/addTaskForm/{employeeId}")
    public String addTaskForm(@PathVariable("employeeId") int employeeId, Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }

        if (!session.isAdmin()) {
            return "AccessDenied";
        }
        TaskDto theTask = new TaskDto();
        theTask.setEmployeeId(employeeId);

        EmployeeDto theEmployee = employeeService.getEmployee(employeeId);
        theModel.addAttribute("taskOwnerUserName", theEmployee.getUserName());
        theModel.addAttribute("task", theTask);
        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("admin", session.isAdmin());

        return "TaskForm";
    }

    @GetMapping("/addTeamForm")
    public String addTeamForm(Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (!session.isAdmin()) {
            return "AccessDenied";
        }
        TeamDto team = new TeamDto();

        theModel.addAttribute("team", team);
        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("admin", session.isAdmin());

        theModel.addAttribute("employeesNotInTeam", employeeService.getEmployees());

        return "TeamForm";
    }

    @GetMapping("/deleteTask/{taskId}")
    public String deleteTaskForm(@PathVariable("taskId") int taskId, Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (!session.isAdmin()) {
            return "AccessDenied";
        }
        taskService.deleteTask(Integer.toString(taskId));
        return "redirect:/home";

    }

    @GetMapping("/updateTaskForm/{taskId}")
    public String updateTaskForm(@PathVariable("taskId") int taskId, Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }

        TaskDto theTask = taskService.getTask(taskId);

        if (!(session.isAdmin() || session.getUser().getId() == theTask.getEmployeeId())) {
            return "AccessDenied";
        }

        EmployeeDto theEmployee = employeeService.getEmployee(theTask.getEmployeeId());

        theModel.addAttribute("taskOwnerUserName", theEmployee.getUserName());

        theModel.addAttribute("task", theTask);
        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("admin", session.isAdmin());

        return "TaskForm";
    }

    @GetMapping("/updateTeamForm/{teamId}")
    public String updateTeamForm(@PathVariable("teamId") int teamId, Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }

        if (!session.isAdmin()) {
            return "AccessDenied";
        }
        TeamDto theTeam = teamService.getTeam(teamId);
        List<EmployeeDto> notInTeam = employeeService.getEmployees();

        for (EmployeeDto employeeInTeam : theTeam.getEmployees()) {
            for (EmployeeDto employee : notInTeam) {
                if (employee.getId() == employeeInTeam.getId()) {
                    notInTeam.remove(employee);
                    break;
                }
            }
        }

        theModel.addAttribute("team", theTeam);
        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("admin", session.isAdmin());
        theModel.addAttribute("employeesNotInTeam", notInTeam);

        return "TeamForm";
    }

    @PostMapping("/saveTask")
    public String saveTask(@ModelAttribute("task") TaskDto theTaskDto, Model theModel) {

        if (session == null) {
            return "redirect:/showLoginForm";
        }

        if (session.isAdmin()) {
            if (theTaskDto.getId() == 0) {
                if (taskService.addTask(theTaskDto) == null) {
                    theModel.addAttribute("error", true);
                    return addTaskForm(theTaskDto.getEmployeeId(), theModel);
                }
            } else
                taskService.updateTaskAdmin(theTaskDto);
        } else if (session.getUser().getId() == theTaskDto.getEmployeeId()) {
            taskService.updateTask(theTaskDto);
        } else {
            return "AccessDenied";
        }

        return getTasks(theTaskDto.getEmployeeId(), theModel);

    }

    @PostMapping("/saveTeam")
    public String saveTeam(@ModelAttribute("team") TeamDto theTeamDto, Model theModel) {

        if (session == null) {
            return "redirect:/showLoginForm";
        }

        if (session.isAdmin()) {
            if (theTeamDto.getId() == 0) {
                teamService.addTeam(theTeamDto);
            } else {
                theTeamDto.setEmployees(teamService.getTeam(theTeamDto.getId()).getEmployees());
                teamService.updateTeam(theTeamDto);

            }
        } else {
            return "AccessDenied";
        }

        return "redirect:/teams";

    }

    @PostMapping("/addEmployee/{teamId}/{employeeId}")
    public String addEmployeeToTeam(@PathVariable("teamId") int teamId, @PathVariable("employeeId") int employeeId,
                                    Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (session.isAdmin()) {

            TeamDto team = teamService.getTeam(teamId);

            EmployeeDto employeeInTeam = team.getEmployees().stream().filter(e -> e.getId() == employeeId).findFirst().orElse(null);
            if (employeeInTeam == null) {
                EmployeeDto theEmployee = employeeService.getEmployee(employeeId);
                team.getEmployees().add(theEmployee);
                teamService.updateTeam(team);
            }

            return "redirect:/updateTeamForm/" + teamId;
        } else {
            return "AccessDenied";
        }

    }

    @GetMapping("/deleteTeam/{teamId}")
    public String deleteTeamForm(@PathVariable("teamId") String teamId, Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (!session.isAdmin()) {
            return "AccessDenied";
        }
        teamService.deleteTeam(teamId);
        return "redirect:/teams";

    }

    @GetMapping("/removeEmployeeFromTeam/{teamId}/{employeeId}")
    public String removeEmployeeFromTeam(@PathVariable("teamId") int teamId, @PathVariable("employeeId") int employeeId,
                                         Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (session.isAdmin()) {

            TeamDto team = teamService.getTeam(teamId);

            EmployeeDto employeeToRemove = team.getEmployees().stream().filter(e -> e.getId() == employeeId).findFirst().orElse(null);

            if (employeeToRemove != null) {

                team.getEmployees().remove(employeeToRemove);
                teamService.updateTeam(team);
            }

            return "redirect:/updateTeamForm/" + teamId;
        } else {
            return "AccessDenied";
        }

    }

    @GetMapping("/addTaskToTeamForm")
    public String addTaskToTeamForm(Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (!session.isAdmin()) {
            return "AccessDenied";
        }

        TaskDto theTask = (TaskDto) theModel.getAttribute("task");
        if (theTask == null) {
            theTask = new TaskDto();
        }

        theModel.addAttribute("task", theTask);
        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("admin", session.isAdmin());
        theModel.addAttribute("teams", teamService.getTeams());

        return "TaskToTeamForm";
    }

    @PostMapping("assignTaskToTeam/{teamId}")
    public String assignTaskToTeam(@ModelAttribute("task") TaskDto theTaskDto, Model theModel,
                                   @PathVariable("teamId") int teamId) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (!session.isAdmin()) {
            return "AccessDenied";
        }

        TaskDto assignedTask = teamService.assignTaskToTeam(teamId, theTaskDto);
        if (assignedTask != null) {

            return "redirect:/updateTaskForm/" + assignedTask.getId();
        } else {
            theModel.addAttribute("task", theTaskDto);
            theModel.addAttribute("error", true);

            return addTaskToTeamForm(theModel);
        }

    }

    @GetMapping("/randomPopulationForm")
    public String randomPopulationForm(Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (!session.isAdmin()) {
            return "AccessDenied";
        }

        RandomPopulationInputDto input = new RandomPopulationInputDto();

        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("admin", session.isAdmin());
        theModel.addAttribute("input", input);

        return "RandomPopulationForm";
    }

    @PostMapping("/randomPopulation")
    public String randomPopulation(@ModelAttribute("input") RandomPopulationInputDto input, Model theModel) {

        if (session == null) {
            return "redirect:/showLoginForm";
        }

        if (!session.isAdmin()) {
            return "AccessDenied";
        }

        String result = teamService.randomPopulation(input);

        if (result.equals("invalid input")) {
            theModel.addAttribute("error", true);
            return randomPopulationForm(theModel);
        }

        return "redirect:/home";
    }

    @GetMapping("/addEmployeeForm")
    public String addEmployeeForm(Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }

        if (!session.isAdmin()) {
            return "AccessDenied";
        }
        EmployeeDto employee = new EmployeeDto();

        theModel.addAttribute("employee", employee);
        theModel.addAttribute("user", session.getUser());
        theModel.addAttribute("admin", session.isAdmin());

        return "EmployeeForm";
    }

    @PostMapping("/saveEmployee/{admin}")
    public String saveEmployee(@ModelAttribute("employee") EmployeeDto employeeDto, Model theModel,
                               @PathVariable("admin") int admin) {

        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (!session.isAdmin()) {
            return "AccessDenied";
        }

        employeeService.addEmployee(admin, employeeDto);

        return "redirect:/employees";

    }

    @GetMapping("/deleteEmployee/{employeeId}")
    public String deleteEmployee(@PathVariable("employeeId") int employeeId, Model theModel) {
        if (session == null) {
            return "redirect:/showLoginForm";
        }
        if (!session.isAdmin()) {
            return "AccessDenied";
        }
        String result = employeeService.deleteEmployee(String.valueOf(employeeId));
        if(result.contains("logout"))
            return "redirect:/logout";
        return "redirect:/employees";

    }
}

File: src/main/java/com/alten/springboot/taskmanagerclient/controller/TaskManagerClientController.kt
package com.alten.springboot.taskmanagerclient.controller

import com.alten.springboot.taskmanagerclient.model.EmployeeDto
import com.alten.springboot.taskmanagerclient.model.RandomPopulationInputDto
import com.alten.springboot.taskmanagerclient.model.TaskDto
import com.alten.springboot.taskmanagerclient.model.TeamDto
import com.alten.springboot.taskmanagerclient.service.EmployeeService
import com.alten.springboot.taskmanagerclient.service.LoginService
import com.alten.springboot.taskmanagerclient.service.TaskService
import com.alten.springboot.taskmanagerclient.service.TeamService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpHeaders
import org.springframework.stereotype.Controller
import org.springframework.ui.Model
import org.springframework.ui.set
import org.springframework.web.bind.annotation.*
import javax.servlet.http.HttpServletRequest

@Controller
class TaskManagerClientController (@Autowired private val employeeService: EmployeeService,
                                   @Autowired private val taskService: TaskService,
                                   @Autowired  private val loginService: LoginService,
                                   @Autowired private val teamService: TeamService){

    @GetMapping("/home")
    fun sayHello(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["admin"] = CurrentSessionInfo.isAdmin()

        return "Home"
    }

    @GetMapping("/employees")
    fun getEmployees(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["employees"] = employeeService!!.employees
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        return "Employees"
    }

    @GetMapping("/tasks")
    fun getTasks(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["tasks"] = taskService!!.tasks
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        return "Tasks"
    }

    @GetMapping("/teams")
    fun getTeams(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["teams"] = teamService!!.teams
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        return "Teams"
    }

    @GetMapping("/tasks/{employeeId}")
    fun getTasks(@PathVariable("employeeId") employeeId: Int, theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["tasksOwner"] = employeeService!!.getEmployee(employeeId)
        theModel["tasks"] = taskService!!.getTasksByEmployeeId(employeeId)
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        return "TasksByEmployee"
    }

    @GetMapping("/teams/{teamId}")
    fun getEmployeesByTeam(@PathVariable("teamId") teamId: Int, theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        val theTeam = teamService!!.getTeam(teamId)
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["team"] = theTeam
        theModel["employees"] = theTeam.employees
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        return "EmployeesByTeam"
    }

    @GetMapping("/showLoginForm")
    fun showLoginForm(theModel: Model?): String {
        return "Login"
    }

    @PostMapping("/login")
    fun login(@RequestParam("username") username: String?, @RequestParam("password") password: String?,
              theModel: Model): String {
        val headers = HttpHeaders()
        val result = loginService!!.login(username, password, headers)
        return if (result != null) {
            session = CurrentSessionInfo(headers, result)
            "redirect:/home"
        } else {
            theModel["error"] = true
            showLoginForm(theModel)
        }
    }

    @GetMapping("/logout")
    fun logout(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        loginService!!.logout()
        session = null
        theModel["logout"] = true
        return showLoginForm(theModel)
    }

    @GetMapping("/addTaskForm/{employeeId}")
    fun addTaskForm(@PathVariable("employeeId") employeeId: Int, theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        val theTask = TaskDto()
        theTask.employeeId = employeeId
        val theEmployee = employeeService!!.getEmployee(employeeId)
        theModel["taskOwnerUserName"] = theEmployee.userName
        theModel["task"] = theTask
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        return "TaskForm"
    }

    @GetMapping("/addTeamForm")
    fun addTeamForm(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        val team = TeamDto()
        theModel["team"] = team
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        theModel["employeesNotInTeam"] = employeeService!!.employees
        return "TeamForm"
    }

    @GetMapping("/deleteTask/{taskId}")
    fun deleteTaskForm(@PathVariable("taskId") taskId: Int, theModel: Model?, request: HttpServletRequest?): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        taskService!!.deleteTask(Integer.toString(taskId))
        return "redirect:/home"
    }

    @GetMapping("/updateTaskForm/{taskId}")
    fun updateTaskForm(@PathVariable("taskId") taskId: Int, theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        val theTask = taskService!!.getTask(taskId)
        if (!(CurrentSessionInfo.isAdmin() || CurrentSessionInfo.getUser().id == theTask.employeeId)) {
            return "AccessDenied"
        }
        val theEmployee = employeeService!!.getEmployee(theTask.employeeId)
        theModel["taskOwnerUserName"] = theEmployee.userName
        theModel["task"] = theTask
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        return "TaskForm"
    }

    @GetMapping("/updateTeamForm/{teamId}")
    fun updateTeamForm(@PathVariable("teamId") teamId: Int, theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        val theTeam = teamService!!.getTeam(teamId)
        val notInTeam = employeeService!!.employees
        for (employeeInTeam in theTeam.employees) {
            for (employee in notInTeam) {
                if (employee.id == employeeInTeam.id) {
                    notInTeam.remove(employee)
                    break
                }
            }
        }
        theModel["team"] = theTeam
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        theModel["employeesNotInTeam"] = notInTeam
        return "TeamForm"
    }

    @PostMapping("/saveTask")
    fun saveTask(@ModelAttribute("task") theTaskDto: TaskDto, theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (CurrentSessionInfo.isAdmin()) {
            if (theTaskDto.id == 0) {
                if (taskService!!.addTask(theTaskDto) == null) {
                    theModel["error"] = true
                    return addTaskForm(theTaskDto.employeeId, theModel)
                }
            } else taskService!!.updateTaskAdmin(theTaskDto)
        } else if (CurrentSessionInfo.getUser().id == theTaskDto.employeeId) {
            taskService!!.updateTask(theTaskDto)
        } else {
            return "AccessDenied"
        }
        return getTasks(theTaskDto.employeeId, theModel)
    }

    @PostMapping("/saveTeam")
    fun saveTeam(@ModelAttribute("team") theTeamDto: TeamDto, theModel: Model?): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (CurrentSessionInfo.isAdmin()) {
            if (theTeamDto.id == 0) {
                teamService!!.addTeam(theTeamDto)
            } else {
                theTeamDto.employees = teamService!!.getTeam(theTeamDto.id).employees
                teamService.updateTeam(theTeamDto)
            }
        } else {
            return "AccessDenied"
        }
        return "redirect:/teams"
    }

    @PostMapping("/addEmployee/{teamId}/{employeeId}")
    fun addEmployeeToTeam(@PathVariable("teamId") teamId: Int, @PathVariable("employeeId") employeeId: Int,
                          theModel: Model?): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        return if (CurrentSessionInfo.isAdmin()) {
            val team = teamService!!.getTeam(teamId)
            val employeeInTeam = team.employees.stream().filter { e: EmployeeDto -> e.id == employeeId }.findFirst().orElse(null)
            if (employeeInTeam == null) {
                val theEmployee = employeeService!!.getEmployee(employeeId)
                team.employees.add(theEmployee)
                teamService.updateTeam(team)
            }
            "redirect:/updateTeamForm/$teamId"
        } else {
            "AccessDenied"
        }
    }

    @GetMapping("/deleteTeam/{teamId}")
    fun deleteTeamForm(@PathVariable("teamId") teamId: String?, theModel: Model?): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        teamService!!.deleteTeam(teamId)
        return "redirect:/teams"
    }

    @GetMapping("/removeEmployeeFromTeam/{teamId}/{employeeId}")
    fun removeEmployeeFromTeam(@PathVariable("teamId") teamId: Int, @PathVariable("employeeId") employeeId: Int,
                               theModel: Model?): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        return if (CurrentSessionInfo.isAdmin()) {
            val team = teamService!!.getTeam(teamId)
            val employeeToRemove = team.employees.stream().filter { e: EmployeeDto -> e.id == employeeId }.findFirst().orElse(null)
            if (employeeToRemove != null) {
                team.employees.remove(employeeToRemove)
                teamService.updateTeam(team)
            }
            "redirect:/updateTeamForm/$teamId"
        } else {
            "AccessDenied"
        }
    }

    @GetMapping("/addTaskToTeamForm")
    fun addTaskToTeamForm(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        var theTask = theModel.getAttribute("task") as? TaskDto
        if (theTask == null) {
            theTask = TaskDto()
        }
        theModel["task"] = theTask
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        theModel["teams"] = teamService!!.teams
        return "TaskToTeamForm"
    }

    @PostMapping("assignTaskToTeam/{teamId}")
    fun assignTaskToTeam(@ModelAttribute("task") theTaskDto: TaskDto?, theModel: Model,
                         @PathVariable("teamId") teamId: Int): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        val assignedTask = teamService!!.assignTaskToTeam(teamId, theTaskDto)
        return if (assignedTask != null) {
            "redirect:/updateTaskForm/" + assignedTask.id
        } else {

            if (theTaskDto != null) {
                theModel["task"] = theTaskDto
            }
           theModel["error"] = true
           addTaskToTeamForm(theModel)
        }
    }

    @GetMapping("/randomPopulationForm")
    fun randomPopulationForm(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        val input = RandomPopulationInputDto()
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        theModel["input"] = input
        return "RandomPopulationForm"
    }

    @PostMapping("/randomPopulation")
    fun randomPopulation(@ModelAttribute("input") input: RandomPopulationInputDto?, theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        val result = teamService!!.randomPopulation(input)
        if (result == "invalid input") {
            theModel["error"] = true
            return randomPopulationForm(theModel)
        }
        return "redirect:/home"
    }

    @GetMapping("/addEmployeeForm")
    fun addEmployeeForm(theModel: Model): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        val employee = EmployeeDto()
        theModel["employee"] = employee
        theModel["user"] = CurrentSessionInfo.getUser()
        theModel["admin"] = CurrentSessionInfo.isAdmin()
        return "EmployeeForm"
    }

    @PostMapping("/saveEmployee/{admin}")
    fun saveEmployee(@ModelAttribute("employee") employeeDto: EmployeeDto?, theModel: Model?,
                     @PathVariable("admin") admin: Int): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        employeeService!!.addEmployee(admin, employeeDto)
        return "redirect:/employees"
    }

    @GetMapping("/deleteEmployee/{employeeId}")
    fun deleteEmployee(@PathVariable("employeeId") employeeId: Int, theModel: Model?): String {
        if (session == null) {
            return "redirect:/showLoginForm"
        }
        if (!CurrentSessionInfo.isAdmin()) {
            return "AccessDenied"
        }
        val result = employeeService!!.deleteEmployee(employeeId.toString())
        return if (result.contains("logout")) "redirect:/logout" else "redirect:/employees"
    }

    companion object {
        private var session: CurrentSessionInfo? = null
    }
}